### PR TITLE
Refactor device method subscription code

### PR DESCRIPTION
**Pull request by jspaith**
The existing logic to subscribe to device methods for both the LL and convenience core layers is a bit hard to follow.  Very soon we'll have a new class of method subscription going through similar path (PnP Commands) and trying to add a 3rd mechanism to this won't scale.

Cleanup both sets of code as this I think improves generally and will be pretty critical to add anything else here.

Bonus is than no UT were changed in this refactor.

### CODE REVIEW

File: iothub_client/src/iothub_client_core.c [hunk 1]
```diff
@@ -1991,169 +1991,145 @@ IOTHUB_CLIENT_RESULT IoTHubClientCore_GetTwinAsync(IOTHUB_CLIENT_CORE_HANDLE iot
     return result;
 }
 
+static void FreeDeviceMethodContextIfNeeded(IOTHUB_CLIENT_CORE_INSTANCE* iotHubClientInstance)
```
Comment (ericwol-msft) [inline iothub_client/src/iothub_client_core.c:1994]:
```suggestion
static void FreeDeviceMethodContext(IOTHUB_CLIENT_CORE_INSTANCE* iotHubClientInstance)
```

File: iothub_client/src/iothub_client_core.c [hunk 2]
```diff
@@ -1991,169 +1991,145 @@ IOTHUB_CLIENT_RESULT IoTHubClientCore_GetTwinAsync(IOTHUB_CLIENT_CORE_HANDLE iot
...
+{
+    if (iotHubClientInstance->method_user_context)
+    {
+        free(iotHubClientInstance->method_user_context);
+        iotHubClientInstance->method_user_context = NULL;
+    }
+}
+
+static IOTHUB_CLIENT_RESULT AllocateQueueContextForMethodCallback(IOTHUB_CLIENT_CORE_INSTANCE* iotHubClientInstance, void* userContextCallback)
```
Comment (ericwol-msft) [inline iothub_client/src/iothub_client_core.c:2003]:
```suggestion
static IOTHUB_CLIENT_RESULT allocateQueueContextForMethodCallback(IOTHUB_CLIENT_CORE_INSTANCE* iotHubClientInstance, void* userContextCallback)
```

Comment (ericwol-msft) [inline iothub_client/src/iothub_client_core.c:2003]:
```suggestion
static IOTHUB_CLIENT_RESULT allocateQueueContextForMethodCallback(IOTHUB_CLIENT_CORE_INSTANCE* iotHubClientInstance, void* userContextCallback)
```

File: iothub_client/src/iothub_client_core_ll.c [hunk 3]
```diff
@@ -2550,66 +2599,43 @@ IOTHUB_CLIENT_RESULT IoTHubClientCore_LL_SetDeviceMethodCallback(IOTHUB_CLIENT_C
...
-                /*Codes_SRS_IOTHUBCLIENT_LL_02_019: [If parameter messageCallback is NULL then IoTHubClientCore_LL_SetMessageCallback shall call the underlying layer's _Unsubscribe function and return IOTHUB_CLIENT_OK.] */
-                /*Codes_SRS_IOTHUBCLIENT_LL_12_018: [If deviceMethodCallback is NULL, then IoTHubClientCore_LL_SetDeviceMethodCallback shall call the underlying layer's IoTHubTransport_Unsubscribe_DeviceMethod function and return IOTHUB_CLIENT_OK. ] */
-                /*Codes_SRS_IOTHUBCLIENT_LL_12_022: [ Otherwise IoTHubClientCore_LL_SetDeviceMethodCallback shall succeed and return IOTHUB_CLIENT_OK. ]*/
-                handleData->IoTHubTransport_Unsubscribe_DeviceMethod(handleData->deviceHandle);
-                handleData->methodCallback.type = CALLBACK_TYPE_NONE;
-                handleData->methodCallback.callbackSync = NULL;
-                handleData->methodCallback.userContextCallback = NULL;
-                result = IOTHUB_CLIENT_OK;
-            }
+            ;
```
Comment (ericwol-msft) [inline iothub_client/src/iothub_client_core_ll.c:2606]: ??

Comment (ericwol-msft) [reply iothub_client/src/iothub_client_core_ll.c:2606]: set result to error...or at least add a comment

**Review by ericwol-msft — APPROVED** [review]

File: iothub_client/src/iothub_client_core.c (hunk 1 above)
Comment (jspaith) [reply iothub_client/src/iothub_client_core.c:1994]: Changing per suggestion.

File: iothub_client/src/iothub_client_core.c (hunk 2 above)
Comment (jspaith) [reply iothub_client/src/iothub_client_core.c:2003]: Changing per suggestion.

File: iothub_client/src/iothub_client_core_ll.c (hunk 3 above)
Comment (jspaith) [reply iothub_client/src/iothub_client_core_ll.c:2606]: @ericwol-msft It's not showing up great on GitHub UX but I think you're referring to the lines below.  Correct?  If yes I can add a LogError so it's a bit more clear, though the result is set in the `if` block and the function called logs bunches but for consistency/clarity LogError here is fine.  Assuming this is what you meant right?

```c
        if ((result = VerifyMethodCallbackType(handleData, CALLBACK_TYPE_SYNC, unsubscribing)) != IOTHUB_CLIENT_OK)
        {
            ;
        }
```

**Review by danewalton — APPROVED** [review]

File: iothub_client/src/iothub_client_core_ll.c (hunk 3 above)
Comment (jspaith) [reply iothub_client/src/iothub_client_core_ll.c:2606]: Adding LogError and will get updated out momentarily.
